api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -181,10 +183,24 @@ public static Schema assignFreshIds(int schemaId, Schema schema, NextID nextId)
    * @return a structurally identical schema with new ids assigned by the nextId function
    */
   public static Schema assignFreshIds(Schema schema, Schema baseSchema, NextID nextId) {
-    return new Schema(TypeUtil
-        .visit(schema.asStruct(), new AssignFreshIds(schema, baseSchema, nextId))
-        .asNestedType()
-        .fields());
+    Types.StructType freshSchemaStruct = TypeUtil
+            .visit(schema.asStruct(), new AssignFreshIds(schema, baseSchema, nextId))
+            .asStructType();
+    return new Schema(freshSchemaStruct.fields(), refreshIdentifierFields(freshSchemaStruct, schema));
+  }
+
+  /**
+   * Get the identifier fields in the fresh schema based on the identifier fields in the base schema.
+   * @param freshSchema fresh schema
+   * @param baseSchema base schema
+   * @return idnetifier fields in the fresh schema
+   */
+  public static Set<Integer> refreshIdentifierFields(Types.StructType freshSchema, Schema baseSchema) {
+    Map<String, Integer> nameToId = TypeUtil.indexByName(freshSchema);
+    return baseSchema.identifierFieldNames().stream()
+            .map(nameToId::get)
+            .filter(Objects::nonNull)
```

Comment:
We couldn't just filter the field which couldn't locate a correct field id by the given column name,  instead we should throw an NullPointerException to say that the refresh has been failure.

Reply:
I chose that approach for use case like select, but later realized we should not support that, so throwing exception does make more sense, will update.